### PR TITLE
[Java] Modified ClusterClock and ConsensusModuleAgent to better support time travel tests.

### PR DESCRIPTION
**Pull request by kwbc**
ConsensusModuleAgent generally uses ClusterClock.time() for log timestamps and ClusterClock.timeNanos() for internal time-related tasks.
During time travel tests we want adjust/offset the log timestamps only.
Modified ConsensusModuleAgent to use ClusterClock.convertToNanos instead of timeUnit.convert to allow the implementations of ClusterClock to differentiate between timestamps to be used in the log versus internal time-keeping.
Also modified Election to take both the log timestamp and nano time in several methods to avoid converting back from nanoseconds to log time.

Final note: debated about what to call the log-timestamp parameter. Went with 'timestamp', but perhaps 'now' or 'nowLog' would be better?

### CODE REVIEW

File: aeron-cluster/src/main/java/io/aeron/cluster/ConsensusModuleAgent.java [hunk 1]
```diff
@@ -560,10 +562,10 @@ void onChallengeResponse(final long correlationId, final long clusterSessionId,
 
                 if (session.id() == clusterSessionId && session.state() == CHALLENGED)
                 {
-                    final long now = clusterClock.time();
-                    final long nowMs = clusterTimeUnit.toMillis(now);
-                    session.lastActivityNs(clusterTimeUnit.toNanos(now), correlationId);
-                    authenticator.onChallengeResponse(clusterSessionId, encodedCredentials, nowMs);
+                    final long nowNs = clusterClock.timeNanos();
+                    final long nowMillis = NANOSECONDS.toMillis(nowNs);
```
Comment (mikeb01) [inline aeron-cluster/src/main/java/io/aeron/cluster/ConsensusModuleAgent.java:566]: Should be `nowMs`.

File: aeron-cluster/src/test/java/io/aeron/cluster/ElectionTest.java [hunk 2]
```diff
@@ -1294,4 +1294,9 @@ private Election newElection(
 
         return clusterMembers;
     }
+
+    private static void electionDoWork(final Election election, final long time)
```
Comment (mikeb01) [inline aeron-cluster/src/test/java/io/aeron/cluster/ElectionTest.java:1298]: Should probably be `timeNs` for consistency.

**Comment by mikeb01** [comment]
Overall I'm happy with the change (some small comments mentioned).

One question the addition of `ClusterClock::convertToNanos`.  I'm not sure I completely understand the necessity of it, would it be possible to show with a simple test/use case it's value over calling `clusterClock.timeUnit().toNanos()` directly?

> Final note: debated about what to call the log-timestamp parameter. Went with 'timestamp', but perhaps 'now' or 'nowLog' would be better?

I think `timestamp` is a bit generic.  Not sure what the best name should be (`nowClusterUnits` is a bit wordy, `logTimestamp` is another option).

File: aeron-cluster/src/main/java/io/aeron/cluster/ConsensusModuleAgent.java (hunk 1 above)
Comment (kwbc) [reply aeron-cluster/src/main/java/io/aeron/cluster/ConsensusModuleAgent.java:566]: I had initially named it `nowMs` but thought it looked too much like `nowNs`.  Will change it back to nowMs if that is preferred.

**Comment by kwbc** [comment]
> One question the addition of `ClusterClock::convertToNanos`. I'm not sure I completely understand the necessity of it, would it be possible to show with a simple test/use case it's value over calling `clusterClock.timeUnit().toNanos()` directly?

In our time travel implementation of ClusterClock, `time()` returns `epochNanos() + timeTravelOffset`, and `convertToNanos(time)` returns `time - timeTravelOffset` thus converting the adjusted log timestamp back to real wall-clock time for internal time keeping purposes within the consensus module.  Does that make sense?

Would be simpler/clearer if consensus module used ClusterClock for log timestamps (and timers) only, and some other EpochClock for everything else.  But that would impose a performance penalty having to call ClusterClock.time and EpochClock.time in the same duty-cycle.

**Comment by kwbc** [comment]
> I think `timestamp` is a bit generic. Not sure what the best name should be (`nowClusterUnits` is a bit wordy, `logTimestamp` is another option).

I like `logTimestamp`.  `logClusterUnit` is a bit wordy, but more consistent with `nowNs` and only slightly longer.  Anyone else have an opinion?

**Review by tmontgomery — APPROVED** [review]
`logTimestamp` is fine with me. I also am OK with the PR.

**Comment by kwbc** [comment]
Thanks for the code review comments.  I've renamed timestamp parameters to logTimestamp in ConsensusModuleAgent and Election -- like how it is consistent with logPosition, etc.  And other requested parameter renames.

**Comment by mjpt777** [comment]
I'll review this Friday or over the weekend.
